### PR TITLE
rgw: bucket list operation slow down in special scenario

### PR DESCRIPTION
**Pull request by zealot-xzd**
rgw: bucket list operation slow down in special scenario

consider this scenario:
bucke with 512 shards, many sibling directories(700), many objects(8000) under each directory
if cls_filtered is true, list the sibling directories takes 5min.
after modification, just takes 5sec

Fixes: https://tracker.ceph.com/issues/56396
Signed-off-by: zealot <xzd16@sohu.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
I haven't tried to think through the scenario, more detailed explanation of the optimization would be helpful for non-experts in cls_filter

**Comment by zealot-xzd** [comment]
> list the sibling directories takes 5min.

if cls_filter=true, delimiter filtering is done in the OSD/CLS.

consider a bucket(b-01) with 6 shards:

1. list objects with delimiter='/'
2. every shard return directory(just call it directory)

shard-1:
	b,c,d,e
shard-2:
	a,c,d,e
shard-3:
	b,c,d,e
shard-4:
	d,e,f,g
shard-5:
	e,f,g
shard-6:
	f,g,h

init candidates=[a,b,c,d,e,f]

iterate over candidates,
object 'a' come from shard-2, call function next_candidate, shard-2 will advance to end because all objects of shard-2 in candidates, if shard-2 is is_truncated, we will break the iteration, and just get only one object 'a'.

if this happens,
consider bucke with 512 shards, many sibling directories(700), many objects(8000) under each directory.
list the sibling directories becomes awful.

actually, we can get [a,b,c,d,e] meet the conditions at a time.

**Comment by zealot-xzd** [comment]
> I haven't tried to think through the scenario, more detailed explanation of the optimization would be helpful for non-experts in cls_filter


if cls_filter=true, delimiter filtering is done in the OSD/CLS.

consider a bucket(b-01) with 6 shards:

1. list objects with delimiter='/'
2. every shard return directory(just call it directory)

shard-1:
	b,c,d,e
shard-2:
	a,c,d,e
shard-3:
	b,c,d,e
shard-4:
	d,e,f,g
shard-5:
	e,f,g
shard-6:
	f,g,h

init candidates=[a,b,c,d,e,f]

iterate over candidates,
object 'a' come from shard-2, call function next_candidate, shard-2 will advance to end because all objects of shard-2 in candidates, if shard-2 is is_truncated, we will break the iteration, and just get only one object 'a'.

if this happens,
consider bucke with 512 shards, many sibling directories(700), many objects(8000) under each directory.
list the sibling directories becomes awful.

actually, we can get [a,b,c,d,e] meet the conditions

**Comment by ivancich** [comment]
I'm trying to understand your explanation, @zealot-xzd. If I understand, the same object appears in multiple shard, such as "d" appearing in shard-1, shard-2, shard-3, and shard-4. Does "d" represent a bunch of objects with names that begin with "d"?

**Comment by ivancich** [comment]
If "a", "b", and so forth represent subdirectories where there's a "/" delimiter, perhaps you could use something like:

shard-1:
b/cat , c/fox, d/elephant, e/goose
shard-2:
a/hippo, c/dog, d/fish, e/hippo

or something like that.

Thank you, @zealot-xzd! I'm very interested in improvement of efficiency in bucket listing.

**Comment by zealot-xzd** [comment]
> If "a", "b", and so forth represent subdirectories where there's a "/" delimiter, perhaps you could use something like:
> 
> shard-1: b/cat , c/fox, d/elephant, e/goose shard-2: a/hippo, c/dog, d/fish, e/hippo
> 
> or something like that.
> 
> Thank you, @zealot-xzd! I'm very interested in improvement of efficiency in bucket listing.

yes, "a", "b" represent directories with a "/" dlimiter，and under directories there are many object:  b-01/a/*.txt，b-01/b/*.txt

list operation just list directories in bucket b-01.

**Review by ivancich — CHANGES_REQUESTED** [review]
I appreciate you explaining the issue, @zealot-xzd . I now see what you're doing. I think the challenge with your solution is that the an already complicated algorithm becomes even more complicated and difficult to understand for future engineers. I'm wondering whether a much more simple solution would be to change candidates from std::map to std::multimap and allow there to be multiple entries with the same key.

When an item is added to the result, the advancement would remove all entries with a matching key and advance each of those trackers.

What do you think? Your solution is likely a little more efficient. But I think understandability in this case is far more important. And we don't want to pre-maturely optimize.

Generally speaking, moving that code to a function where *so* many parameters need to be passed in, so that you can call it a second time seems a bit unwieldy. I think a lambda would have been a better choice, although I don't think we'll that function if you and I decide to pursue the std::multimap solution.

So I'm very curious as to your honest thought! Again, thank you. This will be a huge improvement in performance!

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -8769,14 +8795,39 @@ int RGWRados::cls_bucket_list_ordered(const DoutPrefixProvider *dpp,
...
       // as we cannot be certain that one of the next entries needs to
       // come from that shard; S3 and swift protocols allow returning
       // fewer than what was requested
       ldpp_dout(dpp, 10) << __PRETTY_FUNCTION__ <<
-	": stopped accumulating results at count=" << count <<
-	", dirent=\"" << dirent.key <<
-	"\", because its shard is untruncated and exhaused" << dendl;
+       ": stopped accumulating results at count=" << count <<
+       ", dirent=\"" << dirent.key <<
+       "\", because its shard: " << tracker.oid_name << " is untruncated and exhaused" << dendl;
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc:8829]: If you do change this, we can fix two issues with this message, which I think I originally wrote. First, it's spelled "exhausted", with a "t". Second, it should be "truncated", without the "un".

**Comment by zealot-xzd** [comment]
@ivancich 
lambda  is a good ideal.
I considered your point and did not adopt it  in some other place for performance reasons.
From an easy to understand, your point is correct.
I re-implemented this optimization, using std::multimap,
The log information is also modified.

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -8763,20 +8763,31 @@ int RGWRados::cls_bucket_list_ordered(const DoutPrefixProvider *dpp,
...
+    candidates.erase(range.first, range.second);
+    for (auto idx : vidx) {
+      auto& tracker_match = results_trackers.at(idx);
+      tracker_match.advance();
+      next_candidate(cct, tracker_match, candidates, idx);
+      if (tracker_match.at_end() && tracker_match.is_truncated()) {
+        need_to_stop = true;
+        break;
+      }
+    }
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc:8781]: I'm OK with this as is. I do think we can combine these two loops and get rid of `vidx`, but that's up to you.

**Review by ivancich — APPROVED** [review]
Thank you for all the incredible work!! This will be a great improvement in listing. I can't wait for this to get backported to earlier versions!

**Comment by ivancich** [comment]
@zealot-xzd Do you have any interest in adding a test that covers this scenario, so in the future we can prove correct behavior?

**Comment by ivancich** [comment]
Teuthology Results:
* https://pulpito.ceph.com/ivancich-2022-08-31_23:19:38-rgw-wip-eric-testing-1-distro-default-smithi/
* https://pulpito.ceph.com/ivancich-2022-08-31_23:19:11-rgw-wip-eric-testing-1-distro-default-smithi/

**Pull request merged**